src/libraries/System.Security.Cryptography/docs/X509CertificateLoader.SecurityDesign.md

```diff
@@ -0,0 +1,335 @@
+While strictness is a virtue, so is compatibility, and it is in the interest of compatibility (not rejecting inputs that are identified as the correct format by `X509Certificate2.GetCertContentType`) that `X509CertificateLoader` functions in the same manner.
+
+That is to say, supposing the byte sequence `30 02 05 00` were a valid representation of a format supported by `X509CertificateLoader` (which it is not), the loader will have identicial behavior for that input and for `30 02 05 00 BE EF`, silently discarding the two trailing bytes.
+
+Callers implementing a protocol or file format where the trailing data is unacceptable will need to perform their own content-length validation.
+
+#### BER, CER, or DER
+
+Each file format uses the binary encoding most appropriate to that format.
+For example, ITU-T X.509 and IETF RFC 3280 both indicate that the `Certificate` data type must always use the ASN.1 Distinguished Encoding Rules (DER) restrictions; therefore `LoadCertificate` requires a DER encoding.
+Alternatively, IETF RFC 7292 describes PKCS#12 PFX as being in the relaxed Basic Encoding Rules (BER) form, so `LoadPkcs12` only requires BER.
+
+Callers implementing a protocol or file format where either Canonical Encoding Rules (CER) or DER are required where the format is specified only as BER will need to perform their own restricted-encoding validation.
+
+### Textual Encoding
+
+For each file format that has a textual encoding, once `X509CertificateLoader` has moved into loading the textual form it does so by treating the input as ASCII and finding the first validly encoded PEM envelope with a label appropriate to the data type.
+
+So, `LoadCertificate` will skip over the PKCS7 content and only try loading the CERTIFICATE content in a payload like
+
+```
+-----BEGIN PKCS7-----
+base64==
+-----END PKCS7-----
+-----BEGIN CERTIFICATE-----
+base64==
+-----END CERTIFICATE-----
+```
+
+The textual encoding loaders will ignore any text outside the relevant envelope, and as they treat the data as 8-bit ASCII rather than UTF-8, there is no notion of "invalid" data outside the envelope.
+
+## PKCS#12 PFX Considerations
+
+IETF RFC 7292 describes the file format for a PKCS#12 PFX.
+This format has a number of visible aspects that increase the cost of work, a few non-obvious aspects that increase the cost of work, and a few items that manifest as "quirks" on Windows.
+The general stance of `X509CertificateLoader` is to allow the caller to specify limits on anything that controls the amount of work that will be done while loading the PKCS#12 PFX, other than the length of the payload itself (which is already available to the caller).
+As most callers will have no real idea as to what these limits should be, the default experience when using this API will be to utilize system defaults.
```

Comment:
> system defaults

Sometimes in this document we use "system" to mean "whatever the OS loader does", but here it means "ours".